Add `ob upgrade` and migration graph

srid opened the pull request:
- Add `ob upgrade` command to upgrade obelisk of a project and display migrations to be performed
- Add migration graph, backfilled for `master` and this branch

## Status

Functional, and ready for wider testing, however the branch has a trail of TODOs/FIXMEs as potential next stage of work. In order to test, install this branch of Obelisk to your $PATH, then go to your Obelisk project and run `ob upgrade ob-upgrade-pr` (as soon as this PR is merged, we would be able to test it with other open PRs).

## Developer workflow as of now

Immediately after every individual commit they make (in any branch), the developer must generate new vertices/edge for that commit by running the following commands:

```sh
./migration/backfill.sh
git add migration
git commit --amend --no-edit
```

Alternatively they may do it for multiple commits at a time (but not `amend`ing) at the expense of not being able to migrate to/from the parent commit inside.

ryantrinkle commented:
@srid Awesome :) Do you have a sense of how long the TODOs/FIXMEs will take?  I'd definitely like to get this merged ASAP.

Regarding backfill.sh: can they just run it prior to committing, instead of amending?

ryantrinkle commented on lib/command/src/Obelisk/Command/Upgrade.hs:
```diff
@@ -0,0 +1,186 @@
+
+
+ensureCleanProject :: MonadObelisk m => FilePath -> m ()
+ensureCleanProject project = ensureCleanGitRepo project False "Cannot upgrade with uncommited changes"
```
Is there a specific reason we can't upgrade when there are uncommitted changes?  And, if so, why do we not care about ignored changes?

srid replied:
The only files `ob upgrade` will touch in the project repo is the `.obelisk/impl` thunk. 

At minimum we want that path to be clean, but other than that there is no reason to require the rest of the repository to be clean. I just figured it is easier to reuse that function instead of writing another that checks just for the subpath. Should we do the specific check?

ryantrinkle commented on lib/command/src/Obelisk/Command/Upgrade.hs:
```diff
@@ -0,0 +1,186 @@
+decideHandOffToProjectOb project = do
+  ensureCleanProject project
+  updateThunk (toImplDir project) $ \projectOb -> do
+    getMigrationGraph projectOb MigrationGraph_ObeliskHandoff >>= \case
```
I don't think we should be checking the project migration graph for anything.  What we discussed previously was using the ambient ob's migration graph; is there a reason that can't work?

ryantrinkle commented on lib/command/src/Obelisk/Command/Upgrade.hs:
```diff
@@ -0,0 +1,186 @@
+getAmbientOb :: MonadObelisk m => m FilePath
+getAmbientOb = takeDirectory . takeDirectory <$> liftIO getExecutablePath
+
+getEdge :: MonadObelisk m => Migration action -> (Hash, Hash) -> m action
```
We should move all of the non-ob-specific migration graph stuff into a separate module where it can be reused.

ryantrinkle commented on lib/migration/src/Obelisk/Migration.hs:
```diff
@@ -0,0 +1,112 @@
+findPath :: Hash -> Hash -> Migration action -> Maybe [(Hash, Hash)]
+findPath a b (Migration g' _) = fmap listInPairs $ go a b $ adjacencyMap $ getDag g'
+  where
+    -- Do a slow and dummy search. TODO: improve algorithm performance
```
Doesn't the graph lib have an algorithm for this?

srid replied:
No, I couldn't find one. If our implementation gets slower, we may be able to switch over to one of [these algorithms](https://en.wikipedia.org/wiki/Pathfinding#Algorithms).

ryantrinkle commented on lib/migration/src/Obelisk/Migration.hs:
```diff
@@ -0,0 +1,112 @@
+
+-- Return actions that migrate from one hash to another
+runMigration :: Migration action -> Hash -> Hash -> Maybe [(Hash, action)]
+runMigration m a b = fmap getAction <$> findPath a b m
```
All paths in the DAG from `a` to `b` must sum to the same `action` value, otherwise the migration is invalid.  It looks like that check may still need to be added.

srid replied:
Good point; that would also tell us the 'integrity' of the migration graph.

srid replied:
Addressed this in 259b853 and 2e81b11

ryantrinkle commented:
Getting close :)

srid commented:
> @srid Awesome :) Do you have a sense of how long the TODOs/FIXMEs will take? I'd definitely like to get this merged ASAP.

@ryantrinkle Great! I'd say a few hours of work, including addressing your comments above.

> Regarding backfill.sh: can they just run it prior to committing, instead of amending?

Only if we switch it away from using `git ls-tree` to something else for computing hash. AFAIK, `git ls-tree` requires a commit ref (there appears to be no option to make it work for a modified working copy).

There is also the open question of how to ensure, before merging a PR, that we can verify that the branch has a graph edge from latest master to HEAD of the branch. One way is to merge PRs always using `git merge --ff-only` (though I don't think Github UI supports that).

srid commented on lib/cliapp/src/Obelisk/CliApp/Spinner.hs:
```diff
@@ -23,7 +27,15 @@ import Obelisk.CliApp.Types (Cli, CliConfig (..), HasCliConfig, Output (..), get
+-- | Advanced version that controls the last spinner output based on action's result.
+withSpinner'
+  :: (MonadIO m, MonadMask m, Cli m, HasCliConfig m)
+  => (Text, a -> Maybe Text) -- ^ How to (and whether to) render the final spinner message; note that this doesn't take sub-spinner behaviour into account (although it probably should)
```
Improved the API in 6c09b99

srid commented:
This is superseded by #93
